File: java/Trees/PrefixTree.java
package Trees;

import java.util.*;

public class PrefixTree {
    private class TrieNode {
        Map<Character, TrieNode> map = new HashMap<>();
        Boolean isWordUntil = false;
    }

    private TrieNode root;
    private int count;

    public PrefixTree() {
        root = new TrieNode();
        count = 0;
    }

    public void addWord(String word) {
        var temp = root;
        for (int i = 0; i < word.length(); i++) {
            var letter = word.charAt(i);
            if (!temp.map.containsKey(letter)) temp.map.put(letter, new TrieNode());
            temp = temp.map.get(letter);
        }
        temp.isWordUntil = true;
        count++;
    }

    public void removeWord(String word) {
        var temp = root;
        for (int i = 0; i < word.length(); i++) {
            var letter = word.charAt(i);
            if (!temp.map.containsKey(letter)) return;
            temp = temp.map.get(letter);
        }
        temp.isWordUntil = false;
        count--;
    }

    public boolean containsWord(String word) {
        var temp = root;
        for (int i = 0; i < word.length(); i++) {
            var letter = word.charAt(i);
            if (!temp.map.containsKey(letter)) return false;
            temp = temp.map.get(letter);
        }
        return temp.isWordUntil;
    }

    public boolean hasPrefix(String prefix) {
        var temp = root;
        for (int i = 0; i < prefix.length(); i++) {
            var letter = prefix.charAt(i);
            if (!temp.map.containsKey(letter)) return false;
            temp = temp.map.get(letter);
        }
        return true;
    }


    public int countWords() {
        return count;
    }

    public int countWords(String prefix) {
        return getWordsWithPrefix(prefix).size();
    }

    public List<String> getWordsWithPrefix(String prefix) {
        List<String> ans = new LinkedList<>();
        var temp = root;
        for (int i = 0; i < prefix.length(); i++) {
            var letter = prefix.charAt(i);
            if (!temp.map.containsKey(letter)) return ans;
            temp = temp.map.get(letter);
        }
        if (temp.isWordUntil) ans.add(prefix);
        temp.map.forEach((k, v) -> {
            helperGetWords(v, prefix + k, ans);
        });
        return ans;
    }

    private void helperGetWords(TrieNode node, String prefix, List<String> ans) {
        if (node.isWordUntil) ans.add(prefix);
        node.map.forEach((k, v) -> {
            helperGetWords(v, prefix + k, ans);
        });
    }

}

File: typescript/src/trees/prefix-tree.ts
class TrieNode {
  isWordUntil: boolean;
  map: Map<string, TrieNode>;
  constructor() {
    this.isWordUntil = false;
    this.map = new Map();
  }
}

export default class PrefixTree {
  private root: TrieNode = new TrieNode();
  private wordCount: number = 0;

  public addWord(word: string) {
    let node = this.root;
    for (const char of word) {
      if (!node.map.has(char)) node.map.set(char, new TrieNode());
      node = node.map.get(char)!;
    }
    this.wordCount++;
    node.isWordUntil = true;
  }

  public removeWord(word: string) {
    let node = this.root;
    for (const char of word) {
      if (!node.map.has(char)) return;
      node = node.map.get(char)!;
    }
    this.wordCount--;
    node.isWordUntil = false;
  }

  public hasWord(word: string) {
    let node = this.root;
    for (const char of word) {
      if (!node.map.has(char)) return false;
      node = node.map.get(char)!;
    }
    return node.isWordUntil;
  }

  public hasPrefix(prefix: string) {
    let node = this.root;
    for (const char of prefix) {
      if (!node.map.has(char)) return false;
      node = node.map.get(char)!;
    }
    return true;
  }

  public count() {
    return this.wordCount;
  }

  public countWordsWithPrefix(prefix: string) {
    return this.getWordsWithPrefix(prefix).length;
  }

  public getWordsWithPrefix(prefix: string) {
    const ans: string[] = [];
    let node = this.root;
    for (const char of prefix) {
      if (!node.map.has(char)) return ans;
      node = node.map.get(char)!;
    }
    if (node.isWordUntil) ans.push(prefix);
    node.map.forEach((v, k) => {
      this.getWordsWithPrefixHelper(prefix + k, v, ans);
    });
    return ans;
  }

  public getWordsWithPrefixHelper(
    prefix: string,
    node: TrieNode,
    ans: string[],
  ) {
    if (node.isWordUntil) ans.push(prefix);
    node.map.forEach((v, k) => {
      this.getWordsWithPrefixHelper(prefix + k, v, ans);
    });
  }
}
